WINC1500 WiFi module support

tbursztyka opened the pull request:
This is a takeover of the PR #60

With proper style applied, all unaccepted part removed, and further improvements on top.
Wifi control interface (via net-mgmt) worked out.

pnndra committed:
drivers/wifi: Add winc1500 WiFi driver

Adding support for WINC1500 WiFi chip.

It introduces the wifi drivers sub-directory.

It provides a Full-MAC for 802.11 and an offloaded network stack as
well. The driver uses Atmel's winc1500 HAL.

Signed-off-by: Dario Pennisi <dario@iptronix.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>
Signed-off-by: Massimiliano Agneni <massimiliano.agneni@iptronix.com>
Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
samples/net: Add a simple sample to run wifi shell module

Enabling relevant hardware, currently only winc1500 on
quark_se_c1000_devboard. But it could be easily extended to other
drivers/boards.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
ext: Add winc1500 driver from Atmel

origin: extracted from Atmel ASF version 3.35.1

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/wifi: Generalize GPIO configuration for the WINC1500 driver

Previous way was cumbersome (based on an old way for cc2520 actually),
so moving towards the more recent and proper way.

This will enable anybody to actually provide gpio configuration for a
winc1500 out of any board specific location.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/wifi: Move configure macros to Kconfig

There is no point having undocumented and local config options when
Kconfig can be used to expose these relevantly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/wifi: Let's go away from SPI legacy API in WINC1500

As legacy API is meant to be removed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/wifi: Remove useless gpio configuration call in winc1500 driver

This call is already made in nm_bus_wrapper along with SPI.
In nm_bus_wrapper however: moving the call out of #ifdefsf as it is
generic call, whatever bus would be choosen.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/wifi: Split case logic into functions in winc1500 driver

This cleans up the callback functions and isolate each part to a
relevant handler to improve readability.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
I still don't get why github shows the commit in the wrong order...

galak commented:
I'm going to close #60 than, since this takes it over.

jukkar commented on include/net/net_if.h:
```diff
@@ -350,6 +350,12 @@ static inline enum net_verdict net_if_recv_data(struct net_if *iface,
 static inline u16_t net_if_get_ll_reserve(struct net_if *iface,
```
This patch could be sent separately as this is not winc1500 related.

tbursztyka replied:
It was part of original PR, plus: this driver is the one making use of that change.

In a way this PR is more an Net OFFLOAD show off than just a wifi driver only.

jukkar replied:
If we are going to remove offload driver from l2, then we could do it right away. No need to wait for the winc1500 driver here. There are no other offloading device available so doing this removal is fine imho.

jukkar replied:
But if you want to keep it part of this set, then that is fine too. It just looked like the first commit could be separated from this set.

pfalcon commented:
Btw, I hope interested parties are aware that WINC1500 supports proper L2 integration mode (Ethernet emulation, as common), so *could* be integrated without NET_OFFLOAD hacks. (Just a FYI note.)

tbursztyka commented:
@pfalcon I noticed it, looking at the HAL. However, that won't fit with future work bringing wifi generic net_mgmt interface. Also, in future, we could make the net stack building only the bare necessary parts when there would be only offloaded interface(s), to reduce the amount of code on host side. Such a thing we could not do if winc1500 appears as an ethernet interface.

SebastianBoe commented on drivers/wifi/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+if(CONFIG_WIFI_WINC1500)
+  zephyr_include_directories(
+    ./
+    $ENV{ZEPHYR_BASE}/ext/hal/atmel/asf/common/components/wifi/winc1500
```
Pretty sure this line can be omitted because we are already adding this path in 

ext/hal/atmel/asf/common/components/wifi/winc1500/CMakeLists.txt

SebastianBoe commented on ext/hal/atmel/asf/common/components/wifi/winc1500/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+zephyr_include_directories(
+  ./
+  $ENV{ZEPHYR_BASE}/include/drivers
+  $ENV{ZEPHYR_BASE}/drivers/wifi
```
This line is redundant with something done in drivers/wifi/CMakeLists.txt and can be removed.

tbursztyka commented:
@SebastianBoe Thanks, indeed works just fine.

Commit:
drivers/wifi: Switch info level to debug level in winc1500 driver

These output are fine while debugging, but useless otherwise.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@SebastianBoe  I applied your changes btw.

Commit:
drivers/wifi: WINC1500 driver is not using WAKE pin

It does not seem to be useful atm. So let's remove it.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

SebastianBoe commented:
@tbursztyka : Please treat my +1 as a 0. I did not mean to affect the merge-ability, only give some feedback.

tbursztyka commented:
@SebastianBoe ok

tbursztyka commented:
I'll remove this TEST patch, which is not meant to get in the tree anyway.

GAnthony commented on drivers/wifi/Kconfig.winc1500:
```diff
@@ -0,0 +1,62 @@
+	help
+	 SSID length, without counting the terminal character '\0'
+
+config WIFI_WINC1500_PSK
```
So would the actual password be in the project's prj.conf file?  My experience placing passwords in source files, is they have a way of accidentally being pushed up in commits.   I've had this problem in the past, so decided to pass the password and SSID on the build command line, and/or place in local environment variables.
Would the method, as used in this PR: #5028, be perhaps safer?

tbursztyka replied:
Current PR is just the bootstrap of winc1500, which code was primarily made by someone else.

Thing is, inserting fixed credential, as pre-provisioning, at build time or via Kconfig is same: it sucks. 

Btw PR #5028 has a flaw, SSID are not character arrays.

Future patches will introduce a proper wifi net mgmt interface, that all wifi drivers will have to follow.
Such interface could be properly used via a generic pre-provisioning solution via some flash somewhere etc...

GAnthony commented on drivers/wifi/wifi_winc1500.c:
```diff
@@ -0,0 +1,924 @@
+	struct net_buf *frag;
+	int ret;
+
+	w1500_data.socket_data[socket].send_cb = cb;
```
Trying to follow the logic... Where does the send_cb get called to report the actual number of bytes sent?  Is it in a different file?

tbursztyka replied:
original PR did not seem to implement handling SOCKET_MSG_SEND in winc1500_socket_cb()

Quite many stuff were required to be fixed to get that PR in an okeyish state.
further fix will come in future patches, there is a lot to test and fix around that driver and NET_OFFLOAD

Commit:
drivers/winc1500: Implement wifi mgmt offloaded API

Scan, connect and disconnect for now.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pfalcon commented:
@tbursztyka: Ping. Approved (hmm), but CI red, and bunch of conflicts.

Commit:
drivers/wifi: Move all winc1500 related code to its own directory

There will be other drivers, and mixing up all these files together will
create a mess so better having a dedicated place for winc1500, at least.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

dbkinder commented on samples/net/wifi/README.rst:
```diff
@@ -0,0 +1,50 @@
+.. _wifi:
```
I'd recommend a more unique name, maybe ``wifi_sample``  (all these document labels must be unique across the document set).

dbkinder commented on samples/net/wifi/README.rst:
```diff
@@ -0,0 +1,50 @@
+Overview
+********
+
+This sample enables testing Wi-Fi drivers for various boards.
```
This sample allows testing Wi-Fi drivers for various boards by
    enabling the Wi-Fi shell module that provides a set of commands:
    scan, connect, and disconnect.  It also enables the net_shell module
    to verify net_if settings.

dbkinder commented on samples/net/wifi/README.rst:
```diff
@@ -0,0 +1,50 @@
+Building and Running
+********************
+
+Make sure the board and the chip you are targeting are supported for
```
How do you make sure of this?  Should this read:

    Verify the board and chip you are targeting provide Wi-Fi support.

dbkinder commented on samples/net/wifi/README.rst:
```diff
@@ -0,0 +1,50 @@
+
+.. code-block:: console
+
+shell> select wifi
```
the contents of this code block need to be indented to the ``c`` in ``code-block`` (add three leading spaces to lines 35-50)

dbkinder reviewed (approved):
LGTM, thanks!

galak commented:
This needs license review by TSC & board as its pulling code into ext/

galak reviewed (changes requested):
This needs license review by TSC & board as its pulling code into ext/

tbursztyka commented:
@galak there are already ASF components into ext/hal. Does that mean that everytime a bit of code goes to ext/hal it needs to go through TSC? (First time I play in that ext field, so just asking)

galak commented:
> @galak there are already ASF components into ext/hal. Does that mean that everytime a bit of code goes to ext/hal it needs to go through TSC? (First time I play in that ext field, so just asking)

Yes, we haven't been great about this, but any new import of code int EXT needs to go through the TSC and board for license review.

Commit:
ext/hal: Add WINC1500 README for non-Apache 2.0 licensed contribution

Atmel ASF is BSD-3-Clause

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@nashif See 13th commit of this PR. (I guess it won't need to be applied though?)

galak commented:
> @nashif See 13th commit of this PR. (I guess it won't need to be applied though?)
> 
> thanks, I submitted this for review.

Any reason not to pull in the 3.39 release of asf?  Also can we add somewhere in the readme where the code we are pull in is coming from inside of ASF since its a pretty large package (ie some comment about grabbing code from ./common/components/wifi/winc1500/drivers + whatever else we need to import).

tbursztyka commented:
> Any reason not to pull in the 3.39 release of asf? Also can we add somewhere in the readme where the code we are pull in is coming from inside of ASF since its a pretty large package (ie some comment about grabbing code from ./common/components/wifi/winc1500/drivers + whatever else we need to import).

ASF is changing of minor version pretty fast (and this export dates already from last December), but the big issue is they change their API minor version after minor version. When I did the switch to 3.35 I had to change all the relevant code coming from legacy PR, which was not a small task. (And the API changes were only name changes, stupid code style changes like "let's use came case names now!" etc... no logic changes at all.) So it works with 3.35, let's just keep it that way.

nashif commented:
this has been approved by the board now.

tbursztyka commented:
@galak  can you unlock the PR?

nashif commented:
@tbursztyka 

> Also can we add somewhere in the readme where the code we are pull in is coming from inside of ASF since its a pretty large package (ie some comment about grabbing code from ./common/components/wifi/winc1500/drivers + whatever else we need to import).

can you address this please?

tbursztyka commented:
updated. Looks like newer version of ASF are again different in were they put code... (for instance 3.37 embeds also all the stupid tiny examples in the winc1500 common directory)
